crates/polars-core/src/chunked_array/cast.rs

```diff
@@ -98,15 +98,23 @@ where
         }
         match data_type {
             #[cfg(feature = "dtype-categorical")]
-            DataType::Categorical(_) => {
+            DataType::Categorical(rev_map) => {
                 polars_ensure!(
                     self.dtype() == &DataType::UInt32,
                     ComputeError: "cannot cast numeric types to 'Categorical'"
                 );
                 // SAFETY
                 // we are guarded by the type system
                 let ca = unsafe { &*(self as *const ChunkedArray<T> as *const UInt32Chunked) };
-                CategoricalChunked::from_global_indices(ca.clone()).map(|ca| ca.into_series())
+
+                match rev_map {
```

Comment:
We should only allow this in `cast_unchecked`. If we can with a rev-map that is not global we should probably panic here as that is an implementation error on our side.

I see that you check the bounds later, but I believe we can replace the `cast` with `cast_unchecked` on the proper locations to get the proper cast behavior without paying the bound check price.